apps/api/src/app/workflows-v2/usecases/generate-preview/generate-preview.usecase.ts

```diff
@@ -309,3 +321,106 @@ class FrameworkError {
   message: string;
   name: string;
 }
+
+function sanitizeControlValuesByOutputSchema(
+  controlValues: Record<string, unknown>,
+  type: StepTypeEnum
+): Record<string, unknown> {
+  const outputSchema = channelStepSchemas[type].output || actionStepSchemas[type].output;
+
+  if (!outputSchema || !controlValues) {
+    return controlValues;
+  }
+
+  const ajv = new Ajv({ allErrors: true });
+  addFormats(ajv);
+  const validate = ajv.compile(outputSchema);
+  const isValid = validate(controlValues);
+  const errors = validate.errors as null | ErrorObject[];
+
+  if (isValid || !errors || errors?.length === 0) {
+    return controlValues;
+  }
+
+  return replaceInvalidControlValues(controlValues, errors);
```

Comment:
@djabarovgeorge, did we try using the `useDefault` option from AJV so that AJV can make the replacement?

> With [option useDefaults](https://ajv.js.org/guide/options.html#options) Ajv will assign values from default keyword in the schemas of properties and items (when it is the array of schemas) to the missing properties and items.

The flow might be a bit different as AJV will first fill in the gaps with defaults and then do the validation, but I think it's more correct to do it that way than validating and assigning defaults later. The latter has the risk of assigning invalid defaults post-validation.

Note also this `in schemas inside anyOf, oneOf and not (see [#42 (opens new window)](https://github.com/ajv-validator/ajv/issues/42))`